File: backend/app/routers/clients.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from app import schemas, models
from app.database import get_db

router = APIRouter()

@router.post("/clients", response_model=schemas.ClientResponse, status_code=status.HTTP_201_CREATED)
async def create_client(
    client_data: schemas.ClientCreate,
    db: Session = Depends(get_db)
):
    """
    Create a new client
    """
    new_client = models.Client(
        status=client_data.status,
        client_type=client_data.client_type,
        title=client_data.title,
        gender=client_data.gender,
        forename=client_data.forename,
        middle_name=client_data.middle_name,
        surname=client_data.surname,
        date_of_birth=client_data.date_of_birth,
        email=client_data.email,
        secondary_email=client_data.secondary_email,
        landline=client_data.landline
    )
    
    db.add(new_client)
    db.commit()
    db.refresh(new_client)
    
    return new_client

@router.get("/clients", response_model=schemas.ClientsListResponse)
async def get_all_clients(
    skip: int = 0,
    limit: int = 100,
    status_filter: Optional[str] = None,
    search: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """
    Get all clients with optional filtering and search
    """
    query = db.query(models.Client)
    
    # Filter by status if provided
    if status_filter and status_filter != "all":
        query = query.filter(models.Client.status == status_filter)
    
    # Search by name if provided
    if search:
        search_term = f"%{search.lower()}%"
        query = query.filter(
            (models.Client.forename.ilike(search_term)) |
            (models.Client.surname.ilike(search_term)) |
            (models.Client.middle_name.ilike(search_term))
        )
    
    clients = query.order_by(models.Client.surname, models.Client.forename).offset(skip).limit(limit).all()
    total = query.count()
    
    return {
        "clients": clients,
        "total": total
    }

@router.get("/clients/{client_id}", response_model=schemas.ClientResponse)
async def get_client(
    client_id: str,
    db: Session = Depends(get_db)
):
    """
    Get a specific client by ID
    """
    client = db.query(models.Client).filter(models.Client.id == client_id).first()
    
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found"
        )
    
    return client

@router.put("/clients/{client_id}", response_model=schemas.ClientResponse)
async def update_client(
    client_id: str,
    client_data: schemas.ClientUpdate,
    db: Session = Depends(get_db)
):
    """
    Update a client
    """
    client = db.query(models.Client).filter(models.Client.id == client_id).first()
    
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found"
        )
    
    # Update only provided fields
    update_data = client_data.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(client, field, value)
    
    db.commit()
    db.refresh(client)
    
    return client

@router.delete("/clients/{client_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_client(
    client_id: str,
    db: Session = Depends(get_db)
):
    """
    Delete a client
    """
    client = db.query(models.Client).filter(models.Client.id == client_id).first()
    
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found"
        )
    
    db.delete(client)
    db.commit()
    
    return None

File: frontend/src/api/clients.ts
const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000/api';

export interface Client {
  id: string;
  status: string;
  client_type: string;
  title?: string;
  gender?: string;
  forename: string;
  middle_name?: string;
  surname: string;
  date_of_birth?: string;
  email?: string;
  secondary_email?: string;
  landline?: string;
  created_at?: string;
  updated_at?: string;
}

export interface ClientCreate {
  status?: string;
  client_type?: string;
  title?: string;
  gender?: string;
  forename: string;
  middle_name?: string;
  surname: string;
  date_of_birth?: string;
  email?: string;
  secondary_email?: string;
  landline?: string;
}

export interface ClientUpdate {
  status?: string;
  client_type?: string;
  title?: string;
  gender?: string;
  forename?: string;
  middle_name?: string;
  surname?: string;
  date_of_birth?: string;
  email?: string;
  secondary_email?: string;
  landline?: string;
}

export interface ClientsListResponse {
  clients: Client[];
  total: number;
}

export interface ErrorResponse {
  message: string;
}

async function getAuthHeaders() {
  const token = localStorage.getItem('auth_token');
  return {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${token}`,
  };
}

export async function createClient(clientData: ClientCreate): Promise<Client> {
  const response = await fetch(`${API_BASE_URL}/clients`, {
    method: 'POST',
    headers: await getAuthHeaders(),
    body: JSON.stringify(clientData),
  });

  const data = await response.json();

  if (!response.ok) {
    const error = data as ErrorResponse;
    throw new Error(error.message || 'Failed to create client.');
  }

  return data as Client;
}

export async function getAllClients(
  statusFilter?: string,
  search?: string
): Promise<ClientsListResponse> {
  const params = new URLSearchParams();
  if (statusFilter && statusFilter !== 'all') params.append('status_filter', statusFilter);
  if (search) params.append('search', search);

  const url = `${API_BASE_URL}/clients${params.toString() ? `?${params.toString()}` : ''}`;
  const response = await fetch(url, {
    method: 'GET',
    headers: await getAuthHeaders(),
  });

  const data = await response.json();

  if (!response.ok) {
    const error = data as ErrorResponse;
    throw new Error(error.message || 'Failed to fetch clients.');
  }

  return data as ClientsListResponse;
}

export async function getClient(clientId: string): Promise<Client> {
  const response = await fetch(`${API_BASE_URL}/clients/${clientId}`, {
    method: 'GET',
    headers: await getAuthHeaders(),
  });

  const data = await response.json();

  if (!response.ok) {
    const error = data as ErrorResponse;
    throw new Error(error.message || 'Failed to fetch client.');
  }

  return data as Client;
}

export async function updateClient(
  clientId: string,
  clientData: ClientUpdate
): Promise<Client> {
  const response = await fetch(`${API_BASE_URL}/clients/${clientId}`, {
    method: 'PUT',
    headers: await getAuthHeaders(),
    body: JSON.stringify(clientData),
  });

  const data = await response.json();

  if (!response.ok) {
    const error = data as ErrorResponse;
    throw new Error(error.message || 'Failed to update client.');
  }

  return data as Client;
}

export async function deleteClient(clientId: string): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/clients/${clientId}`, {
    method: 'DELETE',
    headers: await getAuthHeaders(),
  });

  if (!response.ok) {
    const data = await response.json();
    const error = data as ErrorResponse;
    throw new Error(error.message || 'Failed to delete client.');
  }
}

